The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9723621999/job/26838952420) [(plain)](https://github.com/rust-lang/rust/commit/1b8bc341fdfd9b064878a8dac4c45c72b3f19790/checks/26838952420/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---

3    |
4 LL | #[used]
5    | ^^^^^^^
+ LL | fn foo() {}
+    | ----------- the kind is `function`
7 error: attribute must be applied to a `static` variable
8   --> $DIR/used.rs:7:1

9    |
---
To only update this specific test, also pass `--test-args used.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/used.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/used" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/used/auxiliary"
--- stderr -------------------------------
error: attribute must be applied to a `static` variable
##[error]  --> /checkout/tests/ui/used.rs:4:1
   |
   |
LL | #[used] //~ ERROR attribute must be applied to a `static` variable
LL | fn foo() {}
LL | fn foo() {}
   | ----------- the kind is `function`
error: attribute must be applied to a `static` variable
##[error]  --> /checkout/tests/ui/used.rs:7:1
   |
   |
LL | #[used] //~ ERROR attribute must be applied to a `static` variable
LL | struct Foo {}
   | ------------- the kind is `struct`

error: attribute must be applied to a `static` variable
error: attribute must be applied to a `static` variable
##[error]  --> /checkout/tests/ui/used.rs:10:1
   |
LL | #[used] //~ ERROR attribute must be applied to a `static` variable
LL | trait Bar {}
   | ------------ the kind is `trait`

error: attribute must be applied to a `static` variable
error: attribute must be applied to a `static` variable
##[error]  --> /checkout/tests/ui/used.rs:13:1
   |
LL | #[used] //~ ERROR attribute must be applied to a `static` variable
LL | impl Bar for Foo {}
   | ------------------- the kind is `implementation block`

error: aborting due to 4 previous errors
```

</details>

--- a/compiler/rustc_passes/messages.ftl
+++ b/compiler/rustc_passes/messages.ftl
@@ -786,6 +786,7 @@ passes_used_compiler_linker =
 
 passes_used_static =
     attribute must be applied to a `static` variable
+    .label = but this is a {$target}
 
 passes_useless_assignment =
     useless assignment of {$is_field_assign ->

--- a/compiler/rustc_passes/src/check_attr.rs
+++ b/compiler/rustc_passes/src/check_attr.rs
@@ -278,7 +278,7 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
         }
 
         self.check_repr(attrs, span, target, item, hir_id);
-        self.check_used(attrs, target);
+        self.check_used(attrs, target, span);
     }
 
     fn inline_attr_str_error_with_macro_def(&self, hir_id: HirId, attr: &Attribute, sym: &str) {
@@ -1978,12 +1978,16 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
         }
     }
 
-    fn check_used(&self, attrs: &[Attribute], target: Target) {
+    fn check_used(&self, attrs: &[Attribute], target: Target, target_span: Span) {
         let mut used_linker_span = None;
         let mut used_compiler_span = None;
         for attr in attrs.iter().filter(|attr| attr.has_name(sym::used)) {
             if target != Target::Static {
-                self.dcx().emit_err(errors::UsedStatic { span: attr.span });
+                self.dcx().emit_err(errors::UsedStatic {
+                    attr_span: attr.span,
+                    span: target_span,
+                    target: target.name(),
+                });
             }
             let inner = attr.meta_item_list();
             match inner.as_deref() {

--- a/compiler/rustc_passes/src/errors.rs
+++ b/compiler/rustc_passes/src/errors.rs
@@ -563,7 +563,10 @@ pub struct ReprConflictingLint;
 #[diag(passes_used_static)]
 pub struct UsedStatic {
     #[primary_span]
+    pub attr_span: Span,
+    #[label]
     pub span: Span,
+    pub target: &'static str,
 }
 
 #[derive(Diagnostic)]

--- a/tests/ui/attributes/used-issue-126789.rs
+++ b/tests/ui/attributes/used-issue-126789.rs
@@ -1,0 +1,6 @@
+extern "C" {
+    #[used] //~ ERROR attribute must be applied to a `static` variable
+    static FOO: i32;
+}
+
+fn main() {}

--- a/tests/ui/attributes/used-issue-126789.stderr
+++ b/tests/ui/attributes/used-issue-126789.stderr
@@ -1,0 +1,10 @@
+error: attribute must be applied to a `static` variable
+  --> $DIR/used-issue-126789.rs:2:5
+   |
+LL |     #[used]
+   |     ^^^^^^^
+LL |     static FOO: i32;
+   |     ---------------- but this is a foreign static item
+
+error: aborting due to 1 previous error
+

--- a/tests/ui/used.stderr
+++ b/tests/ui/used.stderr
@@ -3,24 +3,32 @@ error: attribute must be applied to a `static` variable
    |
 LL | #[used]
    | ^^^^^^^
+LL | fn foo() {}
+   | ----------- but this is a function
 
 error: attribute must be applied to a `static` variable
   --> $DIR/used.rs:7:1
    |
 LL | #[used]
    | ^^^^^^^
+LL | struct Foo {}
+   | ------------- but this is a struct
 
 error: attribute must be applied to a `static` variable
   --> $DIR/used.rs:10:1
    |
 LL | #[used]
    | ^^^^^^^
+LL | trait Bar {}
+   | ------------ but this is a trait
 
 error: attribute must be applied to a `static` variable
   --> $DIR/used.rs:13:1
    |
 LL | #[used]
    | ^^^^^^^
+LL | impl Bar for Foo {}
+   | ------------------- but this is a implementation block
 
 error: aborting due to 4 previous errors
 

